R/tmb-methods.R

```diff
@@ -40,6 +40,46 @@ fitted.mmrm_tmb <- function(object, ...) {
   fitted_col[, 1L, drop = TRUE]
 }
 
+#' @describeIn mmrm_tmb_methods predict conditional means for new data;
+#'  optionally with standard errors and confidence or prediction intervals.
+#'  Returns a vector of predictions if `se.fit == FALSE` and
+#'  `interval == "none"`; otherwise it returns a data.frame with multiple
+#'  columns and one row per input data row.
+#' @importFrom stats predict
+#' @exportS3Method
+#' @examples
+#' stop("implement example")
+predict.mmrm_tmb <- function(
+    object, newdata, se.fit = FALSE,
+    interval = c("none", "confidence"), level = 0.95,
+    na.action = na.pass, ...) {
+  interval <- match.arg(interval)
+  # compute X_new times beta_hat for new data X (conditional mean predictions)
+  x_matrix <- h_get_x_matrix(object, newdata)
+  res <- x_matrix %*% component(object, "beta_est")
+  colnames(res) <- "fit"
+  if (interval != "none" || se.fit) { # compute standard errors of fit
+    se <- sqrt(diag(x_matrix %*% component(object, "beta_vcov") %*% t(x_matrix)))
+    if (se.fit) { # save?
+      res <- cbind(res, se = se)
+    }
+  }
+  if (interval != "none") { # compute confidence interval
+    alpha <- 1 - level
+    # TODO: need to determine df, work with normal for now
+    if (interval == "confidence") {
+      res <- cbind(res,
+          lwr = res[, "fit"] - qnorm(1 - alpha/2) * se,
+          upr = res[, "fit"] + qnorm(1 - alpha/2) * se
+        )
+    }
+  }
+  if (ncol(res) == 1) { # return vector if only fit is computed
```

Comment:
should we return early if only fit is needed?

Reply:
The performance overhead should be minimal (just a bunch of if/elses) and I like to avoid returning early unless there is an error etc.